File: zig/src/perf_test.zig
const std = @import("std");
const commanum = @import("commanum");
const formatWithCommas = commanum.formatWithCommas;
const numberToWords = commanum.numberToWords;
const MAX_LEN = commanum.MAX_LEN;
const WORDS_MAX_LEN = commanum.WORDS_MAX_LEN;

const DEFAULT_NUM_ITERATIONS = 1_000_000;

fn getNumIterations() usize {
    const val = std.posix.getenv("NUM_ITERATIONS") orelse return DEFAULT_NUM_ITERATIONS;
    return std.fmt.parseInt(usize, val, 10) catch DEFAULT_NUM_ITERATIONS;
}

pub fn main() !void {
    const NUM_ITERATIONS = getNumIterations();
    std.debug.print("Zig Performance Test\n", .{});
    std.debug.print("====================\n", .{});
    std.debug.print("Iterations: {d}\n", .{NUM_ITERATIONS});
    std.debug.print("Range: 0 to {d} (max u64)\n\n", .{std.math.maxInt(u64)});

    // Initialize random number generator
    var prng = std.Random.DefaultPrng.init(blk: {
        var seed: u64 = undefined;
        std.posix.getrandom(std.mem.asBytes(&seed)) catch unreachable;
        break :blk seed;
    });
    const random = prng.random();

    const allocator = std.heap.page_allocator;

    // Pre-generate all random number strings
    std.debug.print("Generating {d} random number strings...\n", .{NUM_ITERATIONS});
    var gen_timer = try std.time.Timer.start();

    const num_strs = try allocator.alloc([20]u8, NUM_ITERATIONS);
    defer allocator.free(num_strs);
    const num_lens = try allocator.alloc(u8, NUM_ITERATIONS);
    defer allocator.free(num_lens);

    for (num_strs, num_lens) |*str_buf, *len| {
        const s = std.fmt.bufPrint(str_buf, "{d}", .{random.int(u64)}) catch unreachable;
        len.* = @intCast(s.len);
    }

    const gen_elapsed_ns = gen_timer.read();
    std.debug.print("Generation time: {d:.3} ms\n\n", .{@as(f64, @floatFromInt(gen_elapsed_ns)) / 1_000_000.0});

    var format_buf: [MAX_LEN]u8 = undefined;

    std.debug.print("Running formatting benchmark...\n", .{});

    var bench_timer = try std.time.Timer.start();

    for (num_strs, num_lens) |*str_buf, len| {
        const result = formatWithCommas(str_buf[0..len], &format_buf) catch "";
        std.mem.doNotOptimizeAway(result.ptr);
    }

    const total_elapsed_ns = bench_timer.read();

    // Calculate statistics
    const total_ms: f64 = @as(f64, @floatFromInt(total_elapsed_ns)) / 1_000_000.0;
    const avg_ns: f64 = @as(f64, @floatFromInt(total_elapsed_ns)) / @as(f64, @floatFromInt(NUM_ITERATIONS));
    const ops_per_sec: f64 = @as(f64, @floatFromInt(NUM_ITERATIONS)) / (total_ms / 1000.0);

    // Print results
    std.debug.print("\nformatWithCommas Results:\n", .{});
    std.debug.print("------------------------\n", .{});
    std.debug.print("Total benchmark time:  {d:.3} ms\n", .{total_ms});
    std.debug.print("Average per operation: {d:.1} ns\n", .{avg_ns});
    std.debug.print("Throughput: {d:.0} ops/sec\n", .{ops_per_sec});

    // numberToWords benchmark
    var words_buf: [WORDS_MAX_LEN]u8 = undefined;

    std.debug.print("\nRunning numberToWords benchmark...\n", .{});

    var words_timer = try std.time.Timer.start();

    for (num_strs, num_lens) |*str_buf, len| {
        _ = numberToWords(str_buf[0..len], &words_buf) catch {};
    }

    const words_elapsed_ns = words_timer.read();

    const words_ms: f64 = @as(f64, @floatFromInt(words_elapsed_ns)) / 1_000_000.0;
    const words_avg_ns: f64 = @as(f64, @floatFromInt(words_elapsed_ns)) / @as(f64, @floatFromInt(NUM_ITERATIONS));
    const words_ops: f64 = @as(f64, @floatFromInt(NUM_ITERATIONS)) / (words_ms / 1000.0);

    std.debug.print("\nnumberToWords Results:\n", .{});
    std.debug.print("---------------------\n", .{});
    std.debug.print("Total benchmark time:  {d:.3} ms\n", .{words_ms});
    std.debug.print("Average per operation: {d:.1} ns\n", .{words_avg_ns});
    std.debug.print("Throughput: {d:.0} ops/sec\n", .{words_ops});
}

File: lua/perf_test.lua
#!/usr/bin/env lua

-- Import formatWithCommas from commanum module
local commanum = require("commanum")
local formatWithCommas = commanum.formatWithCommas

local NUM_ITERATIONS = tonumber(os.getenv("NUM_ITERATIONS")) or 1000000
local MAX_U64 = 18446744073709551615  -- 2^64 - 1

-- High-resolution timer (uses os.clock for CPU time)
local function get_time_ns()
    return os.clock() * 1e9
end

-- Generate a random number string (Lua's math.random can't handle full u64 range directly)
-- We generate random digit strings to simulate the full range
local function random_u64_string()
    -- Random length between 1 and 20 digits (max u64 is 20 digits)
    local len = math.random(1, 20)
    local digits = {}

    -- First digit can't be 0 (unless it's the only digit)
    if len == 1 then
        digits[1] = tostring(math.random(0, 9))
    else
        digits[1] = tostring(math.random(1, 9))
        for i = 2, len do
            digits[i] = tostring(math.random(0, 9))
        end
    end

    return table.concat(digits)
end

-- Main performance test
local function main()
    print("Lua Performance Test")
    print("====================")
    print(string.format("Iterations: %d", NUM_ITERATIONS))
    print(string.format("Range: 0 to %s (max u64)\n", tostring(MAX_U64)))

    -- Seed random number generator
    math.randomseed(os.time())

    -- Pre-generate all random number strings
    print(string.format("Generating %d random numbers...", NUM_ITERATIONS))
    local gen_start = get_time_ns()

    local numbers = {}
    for i = 1, NUM_ITERATIONS do
        numbers[i] = random_u64_string()
    end

    local gen_end = get_time_ns()
    local gen_elapsed_ms = (gen_end - gen_start) / 1e6
    print(string.format("Generation time: %.3f ms\n", gen_elapsed_ms))

    print("Running formatting benchmark...")

    local bench_start = get_time_ns()

    for i = 1, NUM_ITERATIONS do
        formatWithCommas(numbers[i])
    end

    local bench_end = get_time_ns()
    local total_elapsed_ns = bench_end - bench_start

    -- Calculate statistics
    local total_ms = total_elapsed_ns / 1e6
    local avg_ns = total_elapsed_ns / NUM_ITERATIONS
    local ops_per_sec = NUM_ITERATIONS / (total_ms / 1000)

    -- Print results
    print("\nResults:")
    print("--------")
    print(string.format("Total benchmark time:  %.3f ms", total_ms))
    print(string.format("Average per operation: %.1f ns", avg_ns))
    print(string.format("Throughput: %.0f ops/sec", ops_per_sec))
end

main()
